src/olympia/amo/indexers.py

```diff
@@ -0,0 +1,147 @@
+import logging
+
+from django.conf import settings
+
+from olympia.constants.search import (
+    SEARCH_ANALYZER_MAP, SEARCH_ANALYZER_PLUGINS)
+
+from .models import SearchMixin
+from .utils import to_language
+
+
+log = logging.getLogger('z.es')
+
+
+class BaseSearchIndexer(object):
+    """
+    Base Indexer class for all search-related things (as opposed to
+    stats-related things).
+
+    Intended to be inherited from every document type that we want to put in
+    ElasticSearch for search-related purposes. A class inheriting from
+    BaseSearchIndexer should implement the following classmethods:
+
+    - get_model(cls)
+    - get_mapping(cls)
+    - extract_document(cls, obj)
+    """
+
+    @classmethod
+    def get_index_alias(cls):
+        """Return the index alias name."""
+        return settings.ES_INDEXES.get(SearchMixin.ES_ALIAS_KEY)
+
+    @classmethod
+    def get_doctype_name(cls):
+        """Return the document type name for this indexer. We default to simply
+        use the db table from the corresponding model."""
+        return cls.get_model()._meta.db_table
+
+    @classmethod
+    def attach_translation_mappings(cls, mapping, field_names):
+        """
+        For each field in field_names, attach a dict to the ES mapping
+        properties making "<field_name>_translations" an object containing
+        "string" and "lang" as non-indexed strings.
+
+        Used to store non-indexed, non-analyzed translations in ES that will be
+        sent back by the API for each item. It does not take care of the
+        indexed content for search, it's there only to store and return
+        raw translations.
+        """
+        doc_name = cls.get_doctype_name()
+
+        for field_name in field_names:
+            # _translations is the suffix in TranslationSerializer.
+            mapping[doc_name]['properties'].update({
+                '%s_translations' % field_name: {
+                    'type': 'object',
+                    'properties': {
+                        'lang': {'type': 'string', 'index': 'no'},
+                        'string': {'type': 'string', 'index': 'no'}
+                    }
+                }
+            })
+
+    @classmethod
+    def attach_language_specific_analyzers(cls, mapping, field_names):
+        """
+        For each field in field_names, attach language-specific mappings that
+        will use specific analyzers for these fields in every language that we
+        support.
+
+        These mappings are used by the search filtering code if they exist.
+        """
+        doc_name = cls.get_doctype_name()
+
+        for analyzer in SEARCH_ANALYZER_MAP:
+            if (not settings.ES_USE_PLUGINS and
+                    analyzer in SEARCH_ANALYZER_PLUGINS):
+                log.info('While creating mapping, skipping the %s analyzer '
```

Comment:
This will go away: I've noticed that we never enabled `ES_USE_PLUGINS` in AMO. More cleanup!